### PR TITLE
[PPML] Create Dockerfile for ppml gramine base image

### PR DESCRIPTION
**Pull request by gc-fu**
### 1. Why the change?
Now nearly all applications using SGX with Gramine use the same image (intelanayltics/bigdl-ppml-trusted-big-data-ml-python-gramine), which can cause the size of the image to become very large.   Besides, each application may only use part of the image.

A better approach should be provide a image for each of the applications.

### 2. For those who wants to extend this image:
Now the working directory is changed from `/ppml/trusted-big-dasta-ml/` to `/ppml/`.
All the corresponding files including
1. bash.manifest.template
2. clean.sh
3. init.sh
4. Makefile

are changed accordingly.

The following four directories required by the `bash.manifest.template` are no longer added to the image.
```bash
  { path = "/root/.kube/", uri = "file:/root/.kube/" },
  { path = "/root/.keras", uri = "file:/root/.keras" },
  { path = "/root/.m2", uri = "file:/root/.m2" },
  { path = "/root/.zinc", uri = "file:/root/.zinc" },
```
For those who wants to use the bash.manifest.template, you need to manually add these four directories into your image.

### 3. Summary of the change 
Provide a base image which contains only Gramine.

Other applications can extend this base image and install their own dependencies.

### 4. How to test?
Run the following command to boot the container:
```bash
export ENCLAVE_KEY=YOUR_ENCLAVE_KEY
export KEYS_PATH=YOUR_KEY_PATH
export SECURE_PASSWORD_PATH=PASSWORD_PATH
export DOCKER_IMAGE=10.239.45.10/arda/intelanayltics/bigdl-ppml-trusted-big-data-ml-python-gramine:gramine-base
sudo docker run -itd \
--privileged \
--net=host \
--name=gc-test-gramine-base \
--cpuset-cpus="20-24" \
--oom-kill-disable \
--device=/dev/sgx/enclave \
--device=/dev/sgx/provision \
-v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket \
-v $ENCLAVE_KEY:/root/.config/gramine/enclave-key.pem \
-v $KEYS_PATH:/ppml/trusted-big-data-ml/work/keys \
-v $SECURE_PASSWORD_PATH:/ppml/trusted-big-data-ml/work/password \
-e SGX_ENABLED=true \
-e SGX_LOG_LEVEL=error \
-e LOCAL_IP=$LOCAL_IP \
$DOCKER_IMAGE
```

Then at `/ppml/trusted-big-data-ml/`, execute the `./init.sh` script.

Verify Gramine is working as follows:

```bash
cd /gramine/CI-Examples/helloworld/
make SGX=1
gramine-sgx helloworld
```

The result should be similar to the following figures:

gramine

Test of the bash.manifest.template:
manifest_test

### 5. TODO List
- [x] Finish Dockerfile
- [x] Solve the TODOs
- [x] Review
- [x] Add README.md
- [x] Remove redundant comments

### CODE REVIEW

File: ppml/base/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ARG no_proxy
+
+FROM ubuntu:20.04
+
+ARG BIGDL_VERSION
+ENV BIGDL_VERSION                       ${BIGDL_VERSION}
+ENV LOCAL_IP                            127.0.0.1
+ENV SGX_MEM_SIZE                        32G
+ENV SGX_LOG_LEVEL                       error
+# TODO: decide whether we need this TINI_VERSION
```
Comment (glorysdj) [inline ppml/base/Dockerfile:13]: TINI is not needed

File: ppml/base/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+# ENV TINI_VERSION                        $TINI_VERSION
+ENV LC_ALL                              C.UTF-8
+ENV LANG                                C.UTF-8
+
+
+
+# TODO: decide whether we need these directories in gramine, or probably move to later images?
```
Comment (glorysdj) [inline ppml/base/Dockerfile:20]: not needed in base image

File: ppml/base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+#     mkdir -p /ppml/trusted-big-data-ml/fl
+
+RUN mkdir -p /ppml/trsuted-big-data-ml/
+
+ADD ./bash.manifest.template /ppml/trusted-big-data-ml/bash.manifest.template
+ADD ./Makefile /ppml/trusted-big-data-ml/Makefile
+ADD ./init.sh /ppml/trusted-big-data-ml/init.sh
+ADD ./clean.sh /ppml/trusted-big-data-ml/clean.sh
+
+# TODO: Three patches, do we really need this?
```
Comment (glorysdj) [inline ppml/base/Dockerfile:42]: I think we need these python patches
if not, the vfork will breaks python code on gramine

Comment (gc-fu) [reply ppml/base/Dockerfile:42]: Thanks, will add it to the Dockerfile

File: ppml/base/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    # zip -u ${BIGDL_HOME}/jars/xgboost4j_2.12-1.1.2.jar /tracker.py && \
+    # patch /usr/local/lib/python3.7/dist-packages/dill/_dill.py /_dill.py.patch && \
+    # patch /usr/lib/python3.7/uuid.py /python-uuid.patch && \
+    # patch /usr/local/lib/python3.7/dist-packages/psutil/_pslinux.py /python-pslinux.patch && \
+    # wget -P /ppml/trusted-big-data-ml/work/lib https://sourceforge.net/projects/analytics-zoo/files/analytics-zoo-data/libhadoop.so && \
+    # cp /usr/lib/x86_64-linux-gnu/libpython3.7m.so.1 /usr/lib/libpython3.7m.so.1 && \
+    chmod a+x /ppml/trusted-big-data-ml/init.sh && \
+    chmod a+x /ppml/trusted-big-data-ml/clean.sh
+
+# TODO: decide whether we need entrypoint here? I guess not
```
Comment (glorysdj) [inline ppml/base/Dockerfile:113]: we dont need entrypoint for k8s

File: ppml/base/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#     mkdir -p /ppml/trusted-big-data-ml/work/models && \
+#     mkdir -p /ppml/trusted-big-data-ml/work/apps && \
+#     mkdir -p /ppml/trusted-big-data-ml/work/examples/bigdl && \
+#     mkdir -p /root/.keras/datasets && \
+#     mkdir -p /root/.zinc && \
+#     mkdir -p /root/.m2 && \
+#     mkdir -p /root/.kube/
+#     mkdir -p /ppml/trusted-big-data-ml/fl
+
+RUN mkdir -p /ppml/trsuted-big-data-ml/
```
Comment (glorysdj) [inline ppml/base/Dockerfile:35]: /ppml

File: ppml/base/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    patch /usr/local/lib/python3.7/dist-packages/dill/_dill.py /_dill.py.patch && \
+    patch /usr/lib/python3.7/uuid.py /python-uuid.patch && \
+    patch /usr/local/lib/python3.7/dist-packages/psutil/_pslinux.py /python-pslinux.patch && \
+    cp /usr/lib/x86_64-linux-gnu/libpython3.7m.so.1 /usr/lib/libpython3.7m.so.1 && \
+    chmod a+x /ppml/init.sh && \
+    chmod a+x /ppml/clean.sh && \
+    chmod +x /ppml/verify-attestation-service.sh
+
+# Install sgxsdk and dcap
+RUN mkdir -p /opt/intel/ && \
```
Comment (glorysdj) [inline ppml/base/Dockerfile:72]: one RUN to run all the commands

File: ppml/base/verify-attestation-service.sh [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    echo "[INFO] PPML Application Exit!"
+    exit 1
+fi
+if [ "$BIGDL_PPML_JAR" = "your_bigdl_ppml_jar" ]; then
+    echo "[ERROR] BIGDL_PPML_JAR is not set!"
+    echo "[INFO] PPML Application Exit!"
+    exit 1
+fi
+
+JARS="$SPARK_HOME/jars/*:$SPARK_HOME/examples/jars/*:$BIGDL_PPML_JAR"
```
Comment (glorysdj) [inline ppml/base/verify-attestation-service.sh:46]: do we need $SPARK_HOME/examples/jars/* in the classpath?

File: ppml/base/verify-attestation-service.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+java -cp $JARS com.intel.analytics.bigdl.ppml.attestation.VerificationCLI -i $APP_ID -k $API_KEY -c $CHALLENGE -u $ATTESTATION_URL -t $ATTESTATION_TYPE
```
Comment (glorysdj) [inline ppml/base/verify-attestation-service.sh:48]: this requires bigdl ppml jar

File: ppml/base/Dockerfile (hunk 6 above)
Comment (gc-fu) [reply ppml/base/Dockerfile:72]: Will do

File: ppml/base/verify-attestation-service.sh (hunk 7 above)
Comment (gc-fu) [reply ppml/base/verify-attestation-service.sh:46]: Will install java8 and required jars in the base image

File: ppml/base/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    wget https://download.01.org/intel-sgx/sgx-dcap/1.13/linux/distro/ubuntu20.04-server/sgx_debian_local_repo.tgz && \
+    tar xzf sgx_debian_local_repo.tgz && \
+    echo 'deb [trusted=yes arch=amd64] file:///opt/intel/sgx_debian_local_repo focal main' | tee /etc/apt/sources.list.d/intel-sgx.list && \
+    wget -qO - https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | apt-key add - && \
+    apt-get update && \
+    apt-get install -y libsgx-enclave-common-dev  libsgx-ae-qe3 libsgx-ae-qve libsgx-urts libsgx-dcap-ql libsgx-dcap-default-qpl libsgx-dcap-quote-verify-dev libsgx-dcap-ql-dev libsgx-dcap-default-qpl-dev libsgx-quote-ex-dev libsgx-uae-service libsgx-ra-network libsgx-ra-uefi
+
+
+
+ENV PYTHONPATH                          /usr/lib/python3.7:/usr/lib/python3.7/lib-dynload:/usr/local/lib/python3.7/dist-packages:/usr/lib/python3/dist-packages
```
Comment (Hanyu-Jin) [inline ppml/base/Dockerfile:88]: The number of spaces seems not right.

Comment (gc-fu) [reply ppml/base/Dockerfile]: Fixed

**Review by Hanyu-Jin — APPROVED** [review]
LGTM

**Review by glorysdj — APPROVED** [review]
LGTM

**Review by ShanSimu — APPROVED** [review]
LGTM

**Pull request merged**